Electrical refactors and performance optimizations

CodeKalias opened the pull request:
### Purpose
Got rid of some string use in WireDirections, ElectrictricityFunctions should run a bit faster and be a little easier to follow, some minor optimizations by replacing .ContainsKey usage and early returns from loops.

### Notes:
There are so many pre-existing comments and functions with no references in ElectricityFunctions, would be nice if they could be trimmed.
See commit message c9fc7ea for an overview of the ElectricityFunctions changes

No obvious issues in testing, though I'd welcome an expert electricsman to verify the results

- [x] Code is sufficiently commented
- [x] Code is indented with tabs and not spaces
- [x] The PR does not include any unnecessary .meta, .prefab or **.unity (scene) changes**
- [x] The PR does not bring up any new compile errors
- [x] The PR has been tested in editor
- [x] The PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)
- [x] The PR has been tested with round restarts.

CodeKalias committed:
Removed some string use from wire sprite code

CodeKalias committed:
ElectricityFunctions.cs refactor, performance should be better

ElectricityFunctions.cs changes

Line 27 public static HashSet<IntrinsicElectronicData> SwitchCaseConnections(Vector2 searchVec,	Matrix matrix, HashSet<PowerTypeCategory> CanConnectTo,	Connection ConnPoints, IntrinsicElectronicData OIinheritance, HashSet<IntrinsicElectronicData> connections)
Order rearranged to allow early return when possible, removed several unnecessary intermediate variables which didn't do anything new or add any readability, whitespace and brackets fixed, reduced amount of casting, handled possibility of undefined behaviour for enum Connection.NA

Line 212 public static float WorkOutAtResistance(ElectronicSupplyData ResistanceSources, IntrinsicElectronicData Indata)
Early return on null, using TryGetValue and caching output and boolean results to reduce unnecessary dictionary access

Line 291 public static float WorkOutResistance(ElectronicSupplyData ResistanceSources, IntrinsicElectronicData Indata)
Using TryGetValue and caching output and boolean results to reduce unnecessary dictionary access, replaced a couple of ifs with a ternary

Line 441 public static float WorkOutVoltageFromConnector(ElectricalOIinheritance ElectricItem, PowerTypeCategory SpecifiedDevice)
Exiting early out of foreach loops on success

Line 478 public static float WorkOutVoltageNOTFromConnector(ElectricalOIinheritance ElectricItem, PowerTypeCategory SpecifiedDevice)
Exiting early out of foreach loops on success

Line 503 public static float WorkOutVoltageFromConnectors(ElectricalOIinheritance ElectricItem, HashSet<PowerTypeCategory> SpecifiedDevices)
Exiting early out of foreach loops on success

CodeKalias committed:
Convert ElectricalNodeControl to TryGetValue instead of ContainsKey and []

CodeKalias committed:
ElectricalSynchronisation TryGetValue's

CodeKalias committed:
More TryGetValues and early breaks

CodeKalias committed:
Reducing levels of indentation, removing duplicate code

Bod9001 commented on UnityProject/Assets/Scripts/Electricity/ConnectionMap.cs:
```diff
@@ -5,191 +5,98 @@ public static class ConnectionMap
+		return (AdjacentConnections.pointA == validDirection || AdjacentConnections.pointB == validDirection);
+	}
 
 	public static bool IsConnectedToTileOverlap(Connection ConnectionDirection, ConnPoint AdjacentConnections)
```
This is meant for overlapping connections  on a tile
like 

![image](https://user-images.githubusercontent.com/27101670/81122673-4c044200-8f29-11ea-8816-1954d1e13e92.png)


Should technically be still connected

Even though to a bug I introduced accidentallyThis doesn't work anymore,  since it expects connections to be a new you Hashset every time, so You have to do record if it got any connections the usual way, if not then check if there is overlap

Bod9001 replied:
It seems to be empty so therefore these wouldn't connect as in it only triggers with machine overlap for some reason? is this an error?

CodeKalias replied:
First comment should be addressed in 1e62f26
Second in 85ab8fd

CodeKalias committed:
Small comment update

More specific explanation of switch fallthrough behavior

CodeKalias committed:
ElectricityFunctions: Don't assume connections is empty

Function SwitchCaseConnections parameter HashSet<IntrinsicElectronicData> connections
May not be empty at the start, changed code to not assume it is.
